File: benchmarks/java/set/Java_Nonpersistent/Sieve.java
import de.hhu.stups.btypes.BSet;
import de.hhu.stups.btypes.BInteger;
import de.hhu.stups.btypes.BBoolean;
import de.hhu.stups.btypes.BUtils;

public class Sieve {






    private BSet<BInteger> numbers;
    private BInteger cur;
    private BInteger limit;

    public Sieve() {
        numbers = BSet.range(new BInteger(2),new BInteger(10000));
        cur = new BInteger(2);
        limit = new BInteger(10000);
    }

    public BInteger ComputeNumberOfPrimes() {
        BInteger res = null;
        while((cur.greater(new BInteger(1)).and(cur.multiply(cur).lessEqual(limit))).booleanValue()) {
            if((numbers.elementOf(cur)).booleanValue()) {
                BInteger n = null;
                BSet<BInteger> set = null;
                n = cur;
                set = new BSet<>();
                while((n.lessEqual(limit.divide(cur))).booleanValue()) {
                    set = set.union(new BSet<>(cur.multiply(n)));
                    n = n.plus(new BInteger(1));
                }
                numbers = numbers.difference(set);
            } 
            cur = cur.plus(new BInteger(1));
        }
        res = numbers.card();
        return res;
    }

    public static void main(String[] args) {
        Sieve exec = new Sieve();
        long start = System.nanoTime();
        exec.ComputeNumberOfPrimes();
        long end = System.nanoTime();
        System.out.println(exec.getClass().toString() + " Execution: " + (end - start));

        Runtime runtime = Runtime.getRuntime();
        long memory = runtime.totalMemory() - runtime.freeMemory();
        System.out.println("Used memory is bytes: " + memory);
        System.out.println("Used memory is megabytes: "
                + (memory / (1024L * 1024L)));
    }


}

File: benchmarks/cpp/set/Sieve.cpp
#include <iostream>
#include <string>
#include "BUtils.cpp"
#include "BSet.cpp"
#include "BInteger.cpp"
#include "BBoolean.cpp"

#ifndef Sieve_H
#define Sieve_H

using namespace std;

class Sieve {



    private:



        BSet<BInteger > numbers;
        BInteger cur;
        BInteger limit;

    public:

        Sieve() {
            numbers = (BSet<BInteger>::range((BInteger(2)),(BInteger(10000))));
            cur = (BInteger(2));
            limit = (BInteger(10000));
        }

        BInteger ComputeNumberOfPrimes() {
            BInteger res;
            while((cur.greater((BInteger(1)))._and(cur.multiply(cur).lessEqual(limit))).booleanValue()) {
                if((numbers.elementOf(cur)).booleanValue()) {
                    BInteger n;
                    BSet<BInteger > set;
                    n = cur;
                    set = (BSet<BInteger >());
                    while((n.lessEqual(limit.divide(cur))).booleanValue()) {
                        set = set._union((BSet<BInteger >(cur.multiply(n))));
                        n = n.plus((BInteger(1)));
                    }
                    numbers = numbers.difference(set);
                } 
                cur = cur.plus((BInteger(1)));
            }
            res = numbers.card();
            return res;
        }

};
int main() {
    clock_t start,finish;
    double time;
    Sieve exec;
    start = clock();
    BInteger result = exec.ComputeNumberOfPrimes();
    finish = clock();
    time = (double(finish)-double(start))/CLOCKS_PER_SEC;
    printf("%f\n", time);
    printf("%d\n", result.intValue());
    return 0;
}
#endif

